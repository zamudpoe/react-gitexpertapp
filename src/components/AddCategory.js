import React, { useState } from 'react'
import PropTypes from "prop-types"

export const AddCategory = ( { setCategories } ) => {
  const [ inputValue, setInputValue ] = useState("")

  const handleInputChange             = ( e ) => { 
    setInputValue(e.target.value) 

    console.log("HandleInputChange llamado");

  }

  const handleSubmit                  = ( e ) => {  
    e.preventDefault() 
    console.log("%chandleSubmit de <AddCategory /> %s","color: coral; font-size: 1.2rem;", inputValue)

    if ( inputValue.trim().length > 2 ) {  
      setCategories( categories => [ inputValue, ...categories ] )
      /* setCategories( 12345 ) */ // forzamos que falle la prueba con esta linea y comentando la de arriba
      setInputValue("")
    } else {  
      console.log("%cPor favor dame mas de dos letras!","color: tomato;")  
    }  
  }

  return (
    <form onSubmit= { handleSubmit }>
      <h1>{ inputValue }</h1>
      <input 
        type        = "text" 
        value       = { inputValue }
        onChange    = { handleInputChange }        
        onSubmit    = { handleSubmit }
        placeholder = "Introduce un valor"
      />
    </form>
  ) 
}


AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

