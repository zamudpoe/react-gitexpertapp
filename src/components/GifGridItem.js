import React from 'react'
import PropTypes from 'prop-types'

export const GifGridItem = ( { id, title, url } ) => {

  /* console.log( { id, title, img } ) */
  /* https://animate.style/ */
  return (
    <div className="card animate__animated animate__zoomInDown" key={ id }>
      <img src={ url } alt={ title } />
      <p> { title } </p>
    </div>
  )

}


/* 
  1. Enzyme
    npm i --save-dev enzyme enzyme-adapter-react-16
  2. Enzyme to JSON
    
  3. PRUEBAS
    *.- Debe moistrar el componente correctamente!.
      *.- Shallow 
      *.- Wrapper
      *- wrapper .toMatchSnapshot()

*/

GifGridItem.propTypes = {
  title : PropTypes.string.isRequired, 
  url   : PropTypes.string.isRequired 
}

