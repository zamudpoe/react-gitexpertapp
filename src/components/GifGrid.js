/* import React, { useState, useEffect, Fragment } from 'react' */
import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem'

export const GifGrid = ( { category } ) => {  
  const { data:images, loading } = useFetchGifs(category)
  /* console.log(data, loading) */

  return ( 

    <>
      <h3>{ category }</h3> <hr />      

      { loading && <p>Loading...</p> }

      {/* 04 destructuramos el estado y lo enviamos al nuevo componente GifGridItem */}
      <div className="card-grid"> 

        {
          images.map( img => ( 
            <GifGridItem 
              key   = { img.id }  
              { ...img }
            />
          ))
        }

      </div>  
    </>

  ) 
}  


