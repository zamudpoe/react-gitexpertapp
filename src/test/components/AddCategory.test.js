import React           from "react";
import "@testing-library/jest-dom";

import { shallow }     from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe('Pruebas en el componente <AddCategory />', () => {

  // Inicializamos variables para todas las pruebas!.  
  const setCategories = jest.fn() ; 
  let wrapper = shallow( <AddCategory setCategories = { setCategories } /> )  

  beforeEach( () => { 
    jest.clearAllMocks()  
    wrapper = shallow( <AddCategory setCategories = { setCategories } /> )  
  } )

  test('Debe de mostrarse correctamente ', () => {
    expect( wrapper ).toMatchSnapshot() 
  })


  test('Debe de cambiar la caja de texto', () => {
    const input  =  wrapper.find('input')
    const value  =  "Hola Mundo"

    // SIMULAMOS el evento "onChange" que cambia el texto del input
    // y le damos el valor de la variable 'value'  
    input.simulate('change', { target : { value } })  

    // COMPARAMOS el valor del input vs valor del 'h1' 
    expect( wrapper.find("h1").text().trim() ).toBe( value )  
  })


  test('No debe de postear la informacion con submit', () => {
    // Cambiar valor de el input para que la prueba no pase!
    /*
    const input  =  wrapper.find('input')
    const value  =  "Hola Mundo"
    input.simulate('change', { target : { value } })   
    */

    // SIMULAMOS el evento submit sin cambiar el valor de la caja de texto!. 
    wrapper.find('form').simulate('submit' , { preventDefault(){} })  
    expect( setCategories ).not.toHaveBeenCalled()  
  })


  test('Debe de llamar el setCategories y limpiar la caja de texto!', () => {

    /*
      1. simular el inputChange
      2. simular el submit 
      3. setCategories se debe de haber llamado!
      4. el input debe de estar vacio!
    */

    // 1. simular el inputChange
    let value    =  "Hola Mundo"
    wrapper.find('input').simulate('change', { target : { value } })  

    // 2. simular el submit 
    wrapper.find('form').simulate('submit' , { preventDefault(){} })  

    // 3. setCategories se debe de haber llamado!
    expect( setCategories ).toHaveBeenCalled()
    expect( setCategories ).toHaveBeenCalledTimes(1) 
    expect( setCategories ).toHaveBeenCalledWith( expect.any( Function ) )

    // 4. el input debe de estar vacio!
    expect( wrapper.find('input').prop( 'value' ) ).toBe( '' )    

  });


});



