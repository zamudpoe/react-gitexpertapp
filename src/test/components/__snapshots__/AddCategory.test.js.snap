// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory /> Debe de mostrarse correctamente  1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <input
    onChange={[Function]}
    onSubmit={[Function]}
    placeholder="Introduce un valor"
    type="text"
    value=""
  />
</form>
`;
